code/game/objects/items/weapons/grenades/gravitygrenade.dm

```diff
@@ -0,0 +1,43 @@
+/obj/item/grenade/gravitygrenade
+	name = "Gravaton Grenade"
+	desc = "This grenade emmits a blast of negative gravatons, inverting gravity in the nearby area for 30 seconds."
+	icon_state = "gravity"
+	item_state = "flashtool"
+	origin_tech = "magnets=4;combat=4"
+	///used if it is a clusterbuster, so it deletes if it fails to activate, and doesnt leave 10+ unactivated nades
+	var/clustersegment = FALSE
+	///How long untill gravity returns (In deciseconds)?
+	var/gravity_timer = 30 SECONDS
+
+/obj/item/grenade/gravitygrenade/prime()
+	update_mob()
+	var/area/area = get_area(src)
+	if(area.outdoors)
+		visible_message("<span class='warning'> The [src] will not work in such a big area!</span>")
+		unprime()
+		return
+	if(!is_teleport_allowed(area.z))
+		visible_message("<span class='warning'> The [src] seems to not work here!</span>")
+		unprime()
+		return
+	if(istype(area, /area/shuttle) || istype(area, /area/ruin))
+		visible_message("<span class='warning'> The [src] seems to not work here!</span>")
+		unprime()
+		return
+	invert_gravity(area)
+	playsound(loc, 'sound/weapons/wave.ogg', 60, 1)
+	addtimer(CALLBACK(null, .proc/invert_gravity, area), gravity_timer)
```

Comment:
Messing with the whole gravity system seems bad to me because it's not really written in a sensible way.

`atom/has_gravity()` despite first checking turf and area stuff will still return true if there's a local gravity gen present.
All while gravity gens try to set the gravity of areas while turning on.

Seems to me there are just better ways to keep track of this. Perhaps removing one of these altogether? This could break stuff, but I have a hard time understanding what the reasoning was behind the current system.

But that's all I could dig up in ten minutes. You can spend some more time looking at this. I guess in the end the maintainers will decide whether this is fine. Code quality is their thing.

Reply:
Adjusted it a bit